Src/Compilers/Core/MSBuildTask/BuildClient.cs

```diff
@@ -0,0 +1,415 @@
+        /// </returns>
+        private static NamedPipeClientStream TryExistingProcesses(
+            string expectedProcessPath,
+            CancellationToken cancellationToken)
+        {
+            Log("Trying existing processes.");
+            foreach (int processId in GetAllProcessIds(expectedProcessPath))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var pipeStream = TryConnectToProcess(
+                    processId,
+                    TimeOutMsExistingProcess,
+                    cancellationToken);
+                if (pipeStream != null)
+                {
+                    Log("Found existing process");
+                    return pipeStream;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// The IsConnected property on named pipes does not detect when the client has disconnected
+        /// if we don't attempt any new I/O after the client disconnects. We start an async I/O here
+        /// which serves to check the pipe for disconnection. 
+        ///
+        /// This will return true if the pipe was disconnected.
+        /// </summary>
+        private static async Task CreateMonitorDisconnectTask(
+            NamedPipeClientStream pipeStream,
+            CancellationToken cancellationToken)
+        {
+            var buffer = new byte[0];
+
+            while (!cancellationToken.IsCancellationRequested && pipeStream.IsConnected)
+            {
+                // Wait a tenth of a second before trying again
+                await Task.Delay(100, cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    Log("Before poking pipe.");
+                    await pipeStream.ReadAsync(buffer, 0, 0, cancellationToken).ConfigureAwait(false);
+                    Log("After poking pipe.");
+                }
+                // Ignore cancellation
+                catch (OperationCanceledException) { }
+                catch (Exception e)
+                {
+                    // It is okay for this call to fail.  Errors will be reflected in the 
+                    // IsConnected property which will be read on the next iteration of the 
+                    LogException(e, "Error poking pipe");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the file path of the executable that started this process.
+        /// </summary>
+        /// <param name="processHandle"></param>
+        /// <param name="flags">Should always be 0: Win32 path format.</param>
+        /// <param name="exeNameBuffer">Buffer for the name</param>
+        /// <param name="bufferSize">
+        /// Size of the buffer coming in, chars written coming out.
+        /// </param>
+        [DllImport("Kernel32.dll", EntryPoint = "QueryFullProcessImageNameW", CharSet = CharSet.Unicode)]
+        static extern bool QueryFullProcessImageName(
```

Comment:
Even with an implementation, I think there are some very Windows-focused assumptions with how we use named pipes. For example, our security claims are based on Windows ACLs on named pipes. Even if it turns out that the exact same things are true of the POSIX named pipe implementation, I still want to verify we aren't making any assumptions.